# Exercícios - aula 01

# Importação das bibliotecas e dos dados

In [22]:
import pandas as pd

In [23]:
# carregando o dataset
df = pd.read_csv('datasets/kc_house_data.csv')

# verificando as 5 primeiras linhas do dataframe
df.head(5)

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

# Limpeza dos dados

In [24]:
# verificando a quantidade de dados duplicados
df['id'].duplicated().value_counts()

False    21436
True       177
Name: id, dtype: int64

In [25]:
# verificando alguns campos duplicados como exemplo
df[df['id'].duplicated() == True].sort_values('id', ascending=True).head(16)

id             date     price  bedrooms  bathrooms  sqft_living  \
2497     1000102  20150422T000000  300000.0         6       3.00         2400   
16815    7200179  20150424T000000  175000.0         2       1.00          840   
11434  109200390  20141020T000000  250000.0         3       1.75         1480   
12418  123039336  20141208T000000  244900.0         1       1.00          620   
7793   251300110  20150114T000000  358000.0         3       2.25         2510   
9235   302000375  20150506T000000  250000.0         3       2.00         1050   
14856  324000530  20150323T000000  459000.0         3       1.00         1320   
7179   526059224  20150206T000000  470000.0         4       1.75         1650   
17382  641900050  20150206T000000  499950.0         4       2.25         2160   
19553  643300040  20150313T000000  719521.0         4       1.75         1920   
15300  705730280  20150421T000000  335000.0         3       2.50         1740   
9276   722039087  20150504T000000  329000.0         2       1.00          990   
3786   723049156  20141112T000000  284700.0         3       1.00         1700   
824    726049190  20150218T000000  431000.0         3       1.00         1810   
17604  795000620  20150311T000000  157000.0         3       1.00         1080   
17603  795000620  20141215T000000  124000.0         3       1.00         1080   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
2497       9373     2.0           0     0  ...      7        2400   
16815     12750     1.0           0     0  ...      6         840   
11434      3900     1.0           0     0  ...      7        1480   
12418      8261     1.0           0     0  ...      5         620   
7793      12013     2.0           0     0  ...      8        2510   
9235      18304     1.0           0     0  ...      7        1050   
14856      5000     1.5           0     0  ...      7        1320   
7179       7276     1.0           0     0  ...      7        1010   
17382      8817     1.0           0     0  ...      7        1460   
19553      9500     1.0           0     0  ...      7        1470   
15300      5267     2.0           0     0  ...      7        1740   
9276      57499     1.0           0     0  ...      6         990   
3786       8645     1.0           0     0  ...      6        1700   
824        7200     1.0           0     0  ...      7        1130   
17604      6250     1.0           0     0  ...      5        1080   
17603      6250     1.0           0     0  ...      5        1080   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
2497               0      1991             0    98002  47.3262 -122.214   
16815              0      1925             0    98055  47.4840 -122.211   
11434              0      1980             0    98023  47.2977 -122.367   
12418              0      1939             0    98106  47.5138 -122.364   
7793               0      1988             0    98003  47.3473 -122.314   
9235               0      1953             0    98001  47.3206 -122.269   
14856              0      1912             0    98116  47.5711 -122.386   
7179             640      1977             0    98011  47.7721 -122.206   
17382            700      1965             0    98133  47.7595 -122.356   
19553            450      1966             0    98006  47.5683 -122.177   
15300              0      1999             0    98038  47.3777 -122.023   
9276               0      1949             0    98070  47.4145 -122.463   
3786               0      1955             0    98146  47.4899 -122.337   
824              680      1954             0    98133  47.7493 -122.351   
17604              0      1950             0    98168  47.5045 -122.330   
17603              0      1950             0    98168  47.5045 -122.330   

       sqft_living15  sqft_lot15  
2497            2060        7316  
16815           1480        6969  
11434           1830        6956  
12418           1180        8244  
7793       

In [26]:
# removendo as linhas duplicadas
df.drop_duplicates(subset='id', keep='last', inplace=True)

# verificando o resultado após a remoção dos dados duplicados
df.duplicated().value_counts()

False    21436
dtype: int64

# Respondendo as questões de negócio

# 1. Quantos imóveis estão disponíveis para compra?

In [27]:
# Verificando o número de linhas
df.shape[0]

21436

R.: Existem 21436 imóveis dispóveis para compra.

# 2. Quantos atributos os imóveis possuem?

In [28]:
# Verificando o número de colunas
df.shape[1]

21

R.: Os imóveis possuem 21 atributos.

# 3. Quais são os atributos?

In [29]:
#3 - Quais são os atributos?
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

R.: Os 21 atributos são: 'id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long', 'sqft_living15', 'sqft_lot15', conforme informado na seção "dicionários das variáveis" no arquivo de introdução.

# 4. Qual o imóvel mais caro do portifólio (imóvel com maior valor)?

In [30]:
# Ordenando o preço em ordem decrescente, mostrando as colunas 'id' e 'price' com o primeiro registro (linha).
df[['id','price']].sort_values('price', ascending=False).head(1)

id      price
7252  6762700020  7700000.0

R.: O imóvel mais caro do portifólio é o de id 6762700020, custando $ 7.700.00,00.

# 5. Qual o imóvel com o maior número de quartos?

In [31]:
# Ordenando por número de quartos em ordem decrescente,mostrando as colunas 'id' e 'bedrooms' com o primeiro registro (linha).
df[['id','bedrooms']].sort_values('bedrooms', ascending=False).head(1)

id  bedrooms
15870  2402100895        33

R.: O imóvel com o maior número de quartos é o de id 2402100895, com 33 quartos.

# 6. Qual a soma total de quartos do conjunto de dados?

In [32]:
# Somando o número total de quartos de todos os imóveis
df['bedrooms'].sum()

72273

R.:  A soma total de quartos do conjunto de dados é 72.273.

# 7. Quantas imóveis possuem 2 banheiros?

In [33]:
# Filtrando casas com dois banheiros e realizando a soma dos imóveis
df.loc[df['bathrooms']==2].shape[0]

1913

R.: 1.913 imóveis possuem dois banheiros.

# 8. Qual o preço médio de todos os imóveis no conjunto de dados?

In [34]:
# Verificando a média dos preços de todos os imóveis
df['price'].mean()

541649.962726255

R.: O preço médio de todos os imóveis do conjunto de dados é $ 541.649,96.

# 9. Qual o preço médio de imóveis com 2 banheiros?

In [35]:
# Filtrando apenas imóveis com 2 banheiros e realizando a média dos preços dos mesmos
df['price'].loc[df['bathrooms']==2].mean()

459307.013591218

R.: O preço médio dos imóveis que possuem 2 banheiros é $ 459.307,01.

# 10. Qual o preço mínimo entre os imóveis com 3 quartos?

In [36]:
# Filtrando apenas imóveis com 3 banheiros e identificando o menor preço entre os mesmos
df['price'].loc[df['bedrooms']==3].min()

89000.0

R.: O preço mínimo entre as casas com 3 quartos é $ 89.000,00.

# 11. Quantos imóveis possuem mais de 300 metros quadrados de sala de estar?

In [37]:
# Criando uma nova coluna com pés convertidos para metros quadrados
df['m2_living'] = df['sqft_living'] * 0.092903

# Filtrando apenas imóveis com mais de 300 metros quadrados e contando quantos imóveis nesta condição
df['id'].loc[df['m2_living'] >300].count()

2251

R.: Existem 2.251 imóveis com mais de 300 metros quadrados de espaço interno.

# 12. Quantos imóveis tem mais de 2 andares?

In [38]:
# Filtrando imóveis com mais de 2 andares e contando quantos imóveis nesta condição
df['id'].loc[df['floors']>2].count()

780

R.: Existem 780 imóveis que possuem mais de 2 andares.

# 13. Quantos imóveis tem vista para o mar?

In [39]:
# Filtrando imóveis que possuem vista para o mar (sinalizados com 1) e contando quantos imóveis nesta condição
df['id'].loc[df['waterfront']==1].count()

163

R.: Existem 163 imóveis que possuem vista para o mar.

# 14. Dos imóveis com vista para o mar, quantos tem 3 quartos?¶

In [40]:
# Filtrando imóveis que possuem vista para o mar (sinalizados com 1), além de possuir mais de 3 quartos e contando quantos imóveis nesta condição
df.loc[(df['waterfront'] == 1) & (df['bedrooms'] == 3)].shape[0]

64

R.: Existem 64 imóveis com vista para o mar e que possuem mais de 3 quartos.

# 15. Dos imóveis com mais de 300 metros quadrados de sala de estar, quantos tem mais de 2 banheiros?

In [41]:
# Filtrando imóveis que possuem mais de 300 metros quadrados e com mais de 2 banheiros, após, contando quantos imóveis nesta condição
df.loc[(df['m2_living']>300) & (df['bathrooms']>2)].shape[0]

2194

R.: Existem 2194 imóveis com mais de 300 metros quadrados e com mais de 2 banheiros.